model_filereading.cpp

```diff
@@ -1,140 +1,180 @@
+
+// split line into substrings when separated by whitespaces
+static inline std::vector<std::string> splitLine(std::string& line){
+  std::istringstream iss(line);
+  std::vector<std::string> substrings((std::istream_iterator<std::string>(iss)), std::istream_iterator<std::string>());
+  return substrings;
+}
+
+// reads radii from a file specified by the filepath and
+// stores them in an unordered map that is an attribute of
+// the Model object
+void Model::readRadiiAndAtomNumFromFile(std::string& filepath){
+  // clear unordered_maps to avoid keeping data from previous runs
+  radii.clear();
+  elem_Z.clear();
+  //TODO if map empty: ask user via dialog box if they want to
+  //TODO reimport the file
+  //TODO make a function "consultUser(string)"
+  //if(radii.empty()){std::cout << "empty" << std::endl;}
+
+  std::string line;
+  std::ifstream inp_file(filepath);
+
+  while(getline(inp_file,line)){
+    std::vector<std::string> substrings = splitLine(line);
+    if(substrings.size() == 3){
+      radii[substrings[1]] = std::stod(substrings[2]);
+      elem_Z[substrings[1]] = std::stoi(substrings[0]);
+    }
+  }
+  return;
+}
+
+// returns the radius of an atom with a given symbol
+inline double Model::findRadiusOfAtom(const std::string& symbol){
+  //TODO add exception handling for when no radius was found:
+  //if(radii[symbol == 0]){
+  //  throw ...;
+  //}
+  //else{ return...;}
+  return radii[symbol];
+}
+
+inline double Model::findRadiusOfAtom(const Atom& at){
+  return findRadiusOfAtom(at.symbol);
+}
+
+void Model::listAtomTypesFromFile(std::string& filepath){
+//if (inp_file.is_open()){  //TODO consider adding an exception, for when file in not valid
+
+  // clear map to avoid keeping data from previous runs
+  number_of_atoms.clear();
+  // we iterate through the lines in the input file
+  std::string line;
+  std::ifstream inp_file(filepath);
+
+  // iterate through lines
+  while(getline(inp_file,line)){ // ws to pass empty lines and blank space before atoms if any
+    // divide line into "words"
+    std::vector<std::string> substrings = splitLine(line);
+      // recognize atom line format: Element_Symbol x y z
+      if (isAtomLine(substrings)) {
+      number_of_atoms[strToValidSymbol(substrings[0])]++;
+    }
+  }
+  // file has been read
+  inp_file.close();
+
+  return;
+}
+
+void Model::readAtomsFromFile(std::string& filepath){
+
+  std::vector<Atom> list_of_atoms;
+
+//if (inp_file.is_open()){  //TODO consider adding an exception, for when file in not valid
+  std::string line;
+  std::ifstream inp_file(filepath);
+
+  // iterate through lines
+  while(getline(inp_file,line)){
+    // divide line into "words"
+    std::vector<std::string> substrings = splitLine(line);
+    // create new atom and add to storage vector if line format corresponds to Element_Symbol x y z
+    if (isAtomLine(substrings)) {
+      std::string valid_symbol = strToValidSymbol(substrings[0]);
+      if (elem_Z[valid_symbol] == NULL){

```

Comment:
You can only compare `NULL` to a pointer. `elem_z` is a unordered map of `int`s, so this is invalid. Does this not cause a compilation error? Are you using `-Werror` as compilation flag?

Reply:
Yes, I saw the error while compiling (but it would still compile). I actually wanted to check if the key existed in the map and I remembered reading that calls to unordered maps with invalid keys returned a NULL value. I might be wrong.

The intend is simply to give atomic number 0 to any custom element symbol not existing in the radii definition file.

Reply:
See [this](http://www.cplusplus.com/reference/unordered_map/unordered_map/count/) for standard solution.